osu.Game.Rulesets.Catch/Objects/Drawables/CaughtObjectContainer.cs

```diff
@@ -0,0 +1,334 @@
+        {
+            entry.LifetimeStart = Clock.CurrentTime;
+
+            lifetimeManager.AddEntry(entry);
+
+            if (entry is StackedObjectEntry stackEntry)
+                addDelayedDropEntry(stackEntry.DelayedDropEntry);
+        }
+
+        private void addDelayedDropEntry(DroppedObjectEntry entry)
+        {
+            entry.LifetimeStart = double.PositiveInfinity;
+
+            lifetimeManager.AddEntry(entry);
+        }
+
+        private bool removeImmediateEntry(CaughtObjectEntry entry)
+        {
+            if (!lifetimeManager.RemoveEntry(entry))
+                return false;
+
+            if (entry is StackedObjectEntry stackEntry)
+                removeDelayedDropEntry(stackEntry.DelayedDropEntry);
+
+            return true;
+        }
+
+        private void removeDelayedDropEntry(DroppedObjectEntry entry)
+        {
+            bool removed = lifetimeManager.RemoveEntry(entry);
+            Debug.Assert(removed);
+        }
+
+        private Vector2 getCurrentDropPosition(Vector2 positionInStack)
+        {
+            return StackedObjectContainer.ToSpaceOfOtherDrawable(positionInStack, droppedObjectTarget);
+        }
+
+        private void dropStackedObject(StackedObjectEntry stackEntry, double time, DroppedObjectAnimation animation, int mirrorDirection)
+        {
+            stackEntry.LifetimeEnd = time;
+
+            var dropEntry = stackEntry.DelayedDropEntry;
+            dropEntry.LifetimeStart = time;
+            dropEntry.Animation = animation;
+            dropEntry.DropPosition = getCurrentDropPosition(stackEntry.PositionInStack);
+            dropEntry.MirrorDirection = mirrorDirection;
+        }
+
+        private void addDrawable(CaughtObjectEntry entry, DrawableCaughtObject drawable)
+        {
+            if (entry is StackedObjectEntry)
+                StackedObjectContainer.Add(drawable);
+            else
+                droppedObjectTarget.Add(drawable);
+
+            drawableMap[entry] = drawable;
+        }
+
+        private void removeDrawable(CaughtObjectEntry entry)
+        {
+            bool removed = drawableMap.Remove(entry, out DrawableCaughtObject drawable);
+            Debug.Assert(removed);
+
+            if (entry is StackedObjectEntry)
+                StackedObjectContainer.Remove(drawable);
+            else
+                droppedObjectTarget.Remove(drawable);
+        }
+
+        private DrawableCaughtObject getPooledDrawable(CatchHitObject hitObject)
+        {
+            switch (hitObject)
+            {
+                case Fruit _:
+                    return caughtFruitPool.Get();
+
+                case Banana _:
+                    return caughtBananaPool.Get();
+
+                case Droplet _:
+                    return caughtDropletPool.Get();
+
+                default:
+                    return null;
+            }
+        }
+
+        private class StackedObjectEntry : CaughtObjectEntry
```

Comment:
Would split these nested classes/enums to separate files just to keep line count of this file down.